File: backend/src/domain/services/DataFlowSampleService.js
'use strict';

/**
 * Deliberate data-flow fixtures: multiple definitions of the same variable,
 * cross-function parameter passing, computational vs. predicate use, and an
 * unreachable-use branch. Exists to give All-Definition/All-Uses coverage
 * analyzers concrete def-use pairs to trace, not to model real behavior.
 */
class DataFlowSampleService {
  /**
   * `total` is defined three times (multiple-definitions handling) and used
   * both computationally (the running sum) and as a predicate (the early-exit
   * check), giving C-Use and P-Use detection something to distinguish.
   */
  static accumulateWithEarlyExit(values, ceiling) {
    let total = 0; // definition #1
    for (const v of values) {
      total = total + v; // definition #2 (redefinition)
      if (total > ceiling) {
        // predicate use of `total`
        total = ceiling; // definition #3
        break;
      }
    }
    return total; // computational use of `total`
  }

  /**
   * Passes a definition across a function boundary (cross-function use), and
   * includes one branch whose assigned value is never read afterward
   * (unreachable-use detection).
   */
  static deriveDiscountTier(orderTotal) {
    const tier = DataFlowSampleService._classifyTier(orderTotal); // cross-function def
    let discountRate = 0;
    if (tier === 'gold') {
      discountRate = 0.15;
    } else if (tier === 'silver') {
      discountRate = 0.08;
    } else {
      discountRate = 0; // reachable but never changes downstream outcome
      return { tier, discountRate, note: 'no discount applied' }; // early return
    }
    const unusedFlag = true; // defined, never used again on this path
    return { tier, discountRate };
  }

  static _classifyTier(orderTotal) {
    if (orderTotal >= 1000) return 'gold';
    if (orderTotal >= 250) return 'silver';
    return 'standard';
  }

  /**
   * Multiple definitions of the same parameter across nested loops, so
   * Partial-Uses-Coverage and Coverage-Reporting-Validation checks have a
   * realistic partially-covered function to measure.
   */
  static buildAdjacencySample(nodeCount) {
    let edgeCount = 0; // definition #1
    const adjacency = [];
    for (let i = 0; i < nodeCount; i++) {
      const row = [];
      for (let j = 0; j < nodeCount; j++) {
        const connected = (i + j) % 3 === 0;
        row.push(connected);
        if (connected) {
          edgeCount = edgeCount + 1; // definition #2, computational use of prior value
        }
      }
      adjacency.push(row);
    }
    return { adjacency, edgeCount };
  }
}

module.exports = DataFlowSampleService;

File: backend/app/domain/services/data_flow_sample_service.py
"""
Deliberate data-flow fixtures: multiple definitions of the same variable,
cross-function parameter passing, computational vs. predicate use, and an
unreachable-use branch. Exists to give All-Definition/All-Uses coverage
analyzers concrete def-use pairs to trace, not to model real behaviour.
"""
from typing import Iterable, Tuple, List, Dict, Any


class DataFlowSampleService:
    @staticmethod
    def accumulate_with_early_exit(values: Iterable[float], ceiling: float) -> float:
        """
        `total` redefined across loop iterations (multiple-definitions handling),
        used both computationally (running sum) and as a predicate (ceiling check).
        """
        total = 0  # definition #1
        for v in values:
            total = total + v  # definition #2
            if total > ceiling:  # predicate use
                total = ceiling  # definition #3
                break
        return total  # computational use

    @staticmethod
    def _classify_tier(order_total: float) -> str:
        if order_total >= 1000:
            return "gold"
        if order_total >= 250:
            return "silver"
        return "standard"

    @staticmethod
    def derive_discount_tier(order_total: float) -> Dict[str, Any]:
        """
        Cross-function def-use: `tier` is defined in _classify_tier and consumed
        here; one branch returns early with a value never read further
        downstream (unreachable-use detection).
        """
        tier = DataFlowSampleService._classify_tier(order_total)  # cross-function definition
        if tier == "gold":
            discount_rate = 0.15
        elif tier == "silver":
            discount_rate = 0.08
        else:
            discount_rate = 0  # reachable, never re-read on this path
            return {"tier": tier, "discountRate": discount_rate, "note": "no discount applied"}
        return {"tier": tier, "discountRate": discount_rate}

    @staticmethod
    def build_adjacency_sample(node_count: int) -> Tuple[List[List[bool]], int]:
        """
        Multiple definitions of `edge_count` across nested loops, so
        Partial-Uses-Coverage and Coverage-Reporting-Validation checks have a
        realistic partially-covered function to measure.
        """
        edge_count = 0  # definition #1
        adjacency: List[List[bool]] = []
        for i in range(node_count):
            row: List[bool] = []
            for j in range(node_count):
                connected = (i + j) % 3 == 0
                row.append(connected)
                if connected:
                    edge_count = edge_count + 1  # definition #2, uses prior value
            adjacency.append(row)
        return adjacency, edge_count
